src/control/server/storage/bdev/backend_vmd.go

```diff
@@ -21,55 +21,76 @@ import (
 	"github.com/daos-stack/daos/src/control/server/storage"
 )
 
-// findPciAddrsWithDomain returns controllers that match the input prefix in the
-// domain component of their PCI address.
-func findPciAddrsWithDomain(inCtrlrs storage.NvmeControllers, prefix string) ([]string, error) {
-	var outPciAddrs []string
-
-	for _, ctrlr := range inCtrlrs {
-		domain, _, _, _, err := common.ParsePCIAddress(ctrlr.PciAddr)
-		if err != nil {
-			return nil, err
+// backingAddrToVMD converts a VMD backing devices address e.g. 5d0505:03.00.0
+// to the relevant logical VMD address e.g. 0000:5d:05.5.
+func backingAddrToVMD(backingAddr string) (string, bool) {
+	domain, _, _, _, err := common.ParsePCIAddress(backingAddr)
+	if err != nil {
+		return "", false
+	}
+	if domain == 0 {
+		return "", false
+	}
+
+	domStr := fmt.Sprintf("%x", domain)
+
+	return fmt.Sprintf("0000:%c%c:%c%c.%c", domStr[0], domStr[1], domStr[2],
+		domStr[3], domStr[5]), true
+}
+
+// mapVMDToBackingAddrs stores found vmd backing addresses under vmd address key.
+func mapVMDToBackingAddrs(foundCtrlrs storage.NvmeControllers) map[string][]string {
+	vmds := make(map[string][]string)
+
+	for _, ctrlr := range foundCtrlrs {
+		// find backing device addresses from vmd address
+		vmdAddr, isVMDBackingAddr := backingAddrToVMD(ctrlr.PciAddr)
+		if isVMDBackingAddr {
+			vmds[vmdAddr] = append(vmds[vmdAddr], ctrlr.PciAddr)
 		}
-		if fmt.Sprintf("%x", domain) == prefix {
-			outPciAddrs = append(outPciAddrs, ctrlr.PciAddr)
+	}
+
+	return vmds
+}
+
+// mapVMDToBackingDevs stores found vmd backing devices under vmd address key.
+func mapVMDToBackingDevs(foundCtrlrs storage.NvmeControllers) map[string]storage.NvmeControllers {
+	vmds := make(map[string]storage.NvmeControllers)
+
+	for _, ctrlr := range foundCtrlrs {
+		// find backing device addresses from vmd address
+		vmdAddr, isVMDBackingAddr := backingAddrToVMD(ctrlr.PciAddr)
+		if isVMDBackingAddr {
+			cn := *ctrlr
+			vmds[vmdAddr] = append(vmds[vmdAddr], &cn)
```

Comment:
I assume this is because the controllers get modified somewhere else, just curious as to where.

Reply:
have moved this function in line to where it gets used, then I'm less concerned about mutation elsewhere and can avoid the copies